---
title: "Derivation of a prognostic signature for renal carcinoma"
output:
  html_document:
    df_print: paged
    toc: true
    toc_depth: 3
    toc_float: false
    number_sections: false
    code_folding: show
    theme: bootstrap
fontsize: 10
editor_options:
  markdown:
    wrap: 72
author: Anna Bird, PhD
---

<br><br>

### Script Overview

This analysis derives a transcriptomic signature that predicts survival
for renal carcinoma patients. Using transcriptomic data from NIH Genomic
Data Commons (GDC), abnormally expressed genes are derived via
differential expression analysis (DEA) of tumor vs. normal associated
tissue (NAT). The resulting deferentially expressed genes (DEG) where
assessed using a Kaplan-Meier survival analysis & Cox Proportional
Hazards (COXPH) model to determine which genes best predict patient
outcome. The resulting prognostic signature shows robust predictive
value of appx. AUC 0.8 & includes a list of genes with well-known
associations to tumor progression, including NFùõãB, DNA repair, & HIF-1‚ç∫
signaling pathways.

<br><br>

**Acronyms**

DEA - Differential expression analysis

DEG - Differentially expressed genes

GDC - Genomic Data Commons (Nation Cancer Institute)

NAT - Normal associated tissue (GDC calls this "Solid tissue normal")

KIRC - Kidney renal clear cell carcinoma

PCA - Principal components analysis

FDR - False discovery rate (adjusted p value)

KM - Kaplan-Meier Analysis

COXPH - Cox Proportional-Hazards model

AUC - Area under the curve; a metric of sensitivity & specificity of a
predictive model.

<br><br>

### Instructions

1)  OPTIONAL: Make sure GDC client is installed on your machine. Default
    setting uses API.
    <https://gdc.cancer.gov/access-data/gdc-data-transfer-tool>

2)  OPTIONAL: Input a file path for your downloaded transcriptomic data
    in the first code chunk below. Total downloaded data are 2.5GB as of
    5/16/22.

3)  Run this script (click "Knit" above) (takes \~15min to run on my
    laptop). If you encounter problems with GDC download, update the
    TCGAbiolinks package and other BioC packages, as described in
    "Preliminaries" Section below.

<br><br>

### Analysis Outline

<br><br>

**Stage 1 - Assess data structure & design Limma model**

1.1) GDC is queried for human Kidney Renal Clear Cell Carcinoma (KIRC)
bulk RNA sequencing of tumor and normal associated tissue (NAT) samples.

Source: <https://portal.gdc.cancer.gov/projects/TCGA-KIRC>

1.2) Clinical metadata are profiled to look for confounders that need to
be corrected for in the limma model design. For example, tumor and NAT
samples showed differences in their tumor stage profile, meaning tumor
stage needed to be corrected for in differential expression analysis
comparing tumor & NAT.

1.3) Library sizes were plotted to determine whether VOOM is needed.
VOOM is typically used for DEA when library size variance is high.
Library size variance is calculated (e.g. determine whether to use
VOOM). This is done by first setting up a preliminary design matrix,
normalizing the data, and plotting the library distributions &
normalization correction factors.

**Stage 2 - Determine which genes show abnormal expression in cancer.
(DEA; tumor vs NAT)**

2.1) Re-run GDC query with a filtered samples list (some samples with
small lib size were removed)

2.2) Limma/VOOM were used to obtain DEG (NAT vs. tumor)

2.3) Determine whether tumor stratification is needed. Asking which
genes significantly predict patient outcomes should take into
consideration whether there are multiple distinct tumor phenotypes. E.g.
if there are two+ tumor sub-types, it may be appropriate to create two+
survival model for each subtype. Correlation heat mapping and PCA were
used to visualize the number of major sample populations & and assess
whether tumor sample stratification might be appropriate (no clear basis
for stratification was observed).

**Stage 3 - Derivation of a prognostic signature using predictive
survival modeling (Kaplan Meier & Cox Prop. Hazards)**

3.1) Create a survival model using the top abnormally expressed genes
(DEG from NAT vs. tumor DEA), were used to determine which genes would
predict survival, yielding a prognostic gene signature.

3.2) Create a Cox Proportional Hazards Model using the prognostic genes
identified in the Kaplan-Meier analysis. The primary COXPH output is an
AUC value over time as an assessment of the overall predictive value of
the prognostic signature genes.

<br><br>

### Preliminaries - Set target directory

Get Started:

Input the file path to which you want to download the GDC data. GDC KIRC
transcriptomic counts will be directed to this folder. Once the
directory has been set, run all code: (click "Run" or "Knit" above).

```{r}
library(rstudioapi)
setwd(dirname(rstudioapi::getActiveDocumentContext()$path)) # setwd to source file location
dir.create("GDCdata")
target_directory <- file.path(getwd(), "GDCdata")   # Optional: change target path for GDC data (~2.5GB)

library(knitr)
knitr::opts_chunk$set(error = FALSE, cache = TRUE, message = FALSE, warning = FALSE, echo = FALSE, include = TRUE, dpi = 300)
```

Even if you have the following BioC packages installed, consider
updating them, as GDC updates their packages frequently.

```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

# Important to have the most updated version of these packs, as GDC is undergoing updates
# BiocManager::install("TCGAbiolinks")
# BiocManager::install("GenomicDataCommons")
# BiocManager::install("BioinformaticsFMRP/TCGAbiolinksGUI.data") 
# BiocManager::install("BioinformaticsFMRP/TCGAbiolinks")     # if this one fails, try the github edition below
# devtools::install_github(repo = "BioinformaticsFMRP/TCGAbiolinks")
```

```{r, message=FALSE, warning= FALSE}
packages <- unique(c("ape", "chroma","colormap","corrplot","cowplot","data.table","datasets","devtools","downloader","Ecdat","embed","epiR","esquisse","extrafont","extrafontdb","factoextra","FactoMineR","flexdashboard","gapminder","ggforce","ggfortify","ggplot2","ggpointdensity","ggpubr","ggrepel","ggsci","ggvis","ggh4x","gmodels","grid","gridBase","gridExtra","gtable","gtools","GGally","heatmaply","hflights","Hmisc","here","hms","jcolors","knitr","kableExtra","Lahman","lattice","lsr","lubridate","Matrix","mlbench","MASS","nycflights13","patchwork","pheatmap","plyr","pzfx","rafalib","RColorBrewer","readxl","remotes","reshape","reshape2","rio","rsvd","scales","sciplot","Seurat","shiny","stringr","stringi","skimr","tidyr","tidyselect","tidyverse","tm","ursa","VennDiagram","viridis","writexl","here","RColorBrewer","repurrrsive","ggridges","ggthemes","tm","SnowballC","wordcloud","e1071","vroom","tidylog","janitor","lubridate","glue","tsibble","tidytext","tidymodels","rstatix","ggplotify","dr4pl","tidyverse","sandwich","lmtest","drc","tidymodels","grid","gridExtra","lattice","parsnip","aod","ggpubr","knitr","patchwork","rlang","rstatix","elisar","ursa","hrbrthemes","ggstatsplot","RPostgreSQL","scclusteval","xlsx","TCGAbiolinks","limma","edgeR","glmnet","factoextra","FactoMineR","caret","SummarizedExperiment","gplots","survival","survminer","RColorBrewer","gProfileR","genefilter", "TCGAbiolinks", "GenomicDataCommons", "CGPfunctions", "survAUC", "caTools", "magick", "webshot"))

package.check <- lapply(
  packages,
  FUN = function(x) {
    if (!require(x, character.only = TRUE)) {
      install.packages(x, dependencies = TRUE)
      library(x, character.only = TRUE)
    }
  }
) 
space_fix <- function(df) {
  df <- df %>% mutate(across(where(is.character), trimws))    
  names(df) <- map(names(df), ~ trimws(.x)) 
  df
}

theme_set(theme_bw(base_family = "Helvetica"))
theme_update(strip.background = element_rect(color="white", fill="#FFFFFF", size=0), 
          strip.text = element_text(colour = 'black', size = 12, face = "bold", hjust = 0),
          plot.tag = element_text(color = "ivory", hjust = 1, size = 8))

theme_set(theme_pubr())

```

<br><br>

### Stage 1 - Assess data structure and plan Limma design matrix

<br><br>

#### 1.1 GDC query

```{r, include = FALSE, echo = FALSE, message = FALSE}
query = GDCquery(
  project = "TCGA-KIRC",
  data.category = "Transcriptome Profiling", 
  experimental.strategy = "RNA-Seq",
  #data.type = "Gene Expression Quantification",
  #workflow.type = "HTSeq - Counts", 
  workflow.type = "STAR - Counts",
  #workflow.type = "IlluminaHiSeq_TotalRNASeqV2",   # Now "" as of 5/14/22    HTSeq - Counts
  sample.type = c("Primary Tumor", "Solid Tissue Normal"),  # tumor and NAT
  access = "open",
  legacy = FALSE)

#TCGAbiolinks:::getProjectSummary("TCGA-KIRC")
GenomicDataCommons::status()

GDCdownload(query,
  method = "api",  
  directory = target_directory,
  files.per.chunk = 10   # only use this with method = "api" 
  )

kirc_res = getResults(query)
```

```{r, echo=FALSE, results='hide', fig.keep='all', message = FALSE}
# Prepare expression series matrix from downloaded files
KIRC_data <- GDCprepare(query, directory = target_directory)
```

```{r}
KIRCMatrix <- assay(KIRC_data)

(t_prefilt_n <- KIRC_data@colData %>% 
  as_tibble(rownames = "samples") %>% 
  dplyr::count(sample_type) %>%
  kbl(caption = "GDC KIRC RNAseq library counts prior to filtering") %>%
  kable_paper("hover", full_width = F))

kableExtra::save_kable(t_prefilt_n, file = "FIGS KIRC prog sig_2/GDC_sampl_nvals_prefilter.html")
```

<br><br>

#### 1.2 Plan DEA based on clinical data

Info on GDC metadata:
<https://docs.gdc.cancer.gov/Data_Dictionary/viewer/#?view=table-definition-view&id=demographic&anchor=age_at_index>

In preparation for the limma differential expression analysis (DEA)
comparing tumor and normal associated tissue (NAT), clinical metadata is
examined here to ask whether several clinical factors may confound the
DEA. For example, is the tumor stage profile or the proportion of males
different between tumor & NAT? Such clinical variables need to be
corrected for statistically in the limma model.

-- Plotting continuous variables --

```{r, fig.width = 5, fig.height = 6, out.width="60%"}

contin_vars <- c("longest_dimension", 
                 "age_at_diagnosis", 
                 "age_at_index")

TTL <- c("Longest dimension (cm)", 
         "Age at diagnosis (days)", 
         "Age at Index (years)")

CV_plot <- function(VAR, ylabel) {

  WT <- tidy(wilcox.test(KIRC_data@colData[[VAR]] ~ definition, KIRC_data@colData))
  WT_p <- WT$p.value
  WT_test <- "wilcoxon"

  df <- KIRC_data@colData %>%
    as_tibble() %>%
    mutate(definition_2 = case_when(definition == "Primary solid Tumor" ~ "Tumor",
                                    definition == "Solid Tissue Normal" ~ "NAT",
                                    TRUE ~ "_"))

  ggplot(df, aes_string(y = VAR, x = factor(df$"definition_2"),
                group = "definition_2",
                fill = "definition_2")) +
    geom_violin(alpha = 0.4, show.legend = F, color = NA) +
    geom_jitter(width = 0.2, alpha = 0.5, show.legend = F) +
    stat_compare_means(method = "wilcox.test",
                       comparisons = list(c(1,2)),
                       aes(label=..p.adj..),
                       label = "p.signif") +
  scale_fill_npg() +
  labs(y = paste(ylabel), x = NULL, 
       title = capitalize(str_replace_all(VAR, "_", " ")), 
       fill = NULL,
       #subtitle = paste("p_val =", round(WT_p, 3), WT_test)
       ) +
  scale_y_continuous(labels = label_number_si())
}


(p_contin_clinvar <- map2(.x = contin_vars, 
                         .y = TTL,
                         ~ CV_plot(.x, .y)) %>%
    gridExtra::marrangeGrob(., 
                            ncol = 2, 
                            nrow = 2,
                            top = NULL
                            ))

dir.create("FIGS KIRC prog sig_2")
ggsave("FIGS KIRC prog sig_2/contin_clin_vars.pdf", 
       p_contin_clinvar, width = 7, height = 7, dpi = 300)

```

Assessment: age_at_index, age_at_diagnosis, and tumor size
(longest_dimension) are not significantly different between primary
solid tumor and solid tissue normal (NAT), & are therefore not
incorporated into the limma design matrix.

<br><br>

Examine metadata for sample variables that would confound a comparison
of tumor & NAT.

-- Plotting discrete variables --

```{r, fig.width = 5, fig.height = 28, out.width="40%"}
discrete_vars <- c("is_ffpe",
                   "gender",
                   "ajcc_pathologic_stage",
                   "prior_treatment",
                   "race",
                   "prior_malignancy")

prop_viewer <- function(VAR) {
  PlotXTabs2(
    data = as_tibble(KIRC_data@colData) %>%
                       mutate(definition_2 = case_when(definition == "Primary solid Tumor" ~ "Tumor",
                                                       definition == "Solid Tissue Normal" ~ "NAT",
                                                       TRUE ~ "_")),
    y = VAR,
    x = definition_2,
    bf.details = TRUE,
    xlab = NULL,
    ylab = NULL,
    data.label = "counts",
    label.fill.alpha = .3,
    legend.position = "left",
    title = paste(VAR, "by tissue"),
    palette = "Set1"
  )   }

map(discrete_vars, ~ prop_viewer(.x)) %>%
  marrangeGrob(., ncol = 1, nrow = 6)


p_ajcc_path <- prop_viewer(discrete_vars[[3]])
ggsave("FIGS KIRC prog sig_2/Limma_clin_adjust ajcc path.tiff", p_ajcc_path, width = 4, height = 4, dpi = 300)

p_gender <- prop_viewer(discrete_vars[[2]])
ggsave("FIGS KIRC prog sig_2/Limma_clin_adjust sex ratio.tiff", p_gender, width = 4, height = 4, dpi = 300)
```

<br><br>

Assessment: ajcc_pathologic_stage is signficantly different between
tumor & NAT, and should be corrected for in the final limma model.
Difference in sex-ratio between tumor vs. NAT will also be corrected for
(even though chi-sq p-val showed differences are not statistically
significant), due to male-female genetic distance.

<br><br>

Create preliminary design matrix comparing sample groups "Solid Tissue
Normal" (NAT) to "Primary solid Tumor"

```{r, results='hide', fig.keep='all', message = FALSE}
targets <- KIRC_data@colData %>% as_tibble %>%
  mutate(ajcc_pathologic_stage = case_when(          # make cancer stage data a continuous & quantitative variable
    ajcc_pathologic_stage == "Stage I" ~ 1,
    ajcc_pathologic_stage == "Stage II" ~ 2,
    ajcc_pathologic_stage == "Stage III" ~ 3,
    ajcc_pathologic_stage == "Stage IV" ~ 4,
    TRUE ~ 2
))  

group <- "definition"
comparison_groups <- KIRC_data@colData$definition %>% unique                 # set comparison groups
design_factor = colData(KIRC_data)[, group, drop=T]                          # choose design factor
design_factor <- factor(
    x = design_factor,
    levels=c("Solid Tissue Normal", "Primary solid Tumor")                   # Set first item here is the reference
)

design <- model.matrix(~ design_factor + gender + ajcc_pathologic_stage, targets)
```

Create DGEList object (S4) that contains counts, norm counts & library
sizes

```{r}
dge = DGEList(counts=assay(KIRC_data),  
              samples=colData(KIRC_data),
              genes=as.data.frame(rowData(KIRC_data)))
```

Remove low expressed genes

```{r}
keep <- filterByExpr(dge, design)       # output is a gene list
dge <- dge[keep,,keep.lib.sizes=FALSE]
dge %>% str(., max.level = 3)
```

<br><br>

#### 1.3 Library QC & filtering

Here, sample library sizes, normalization factors, and expression
distributions are visualized to look for any data abnormalities (i.e.
samples w/ poor sequencing or biological coverage). The red dashed line
shows the library filtering threshold. The removal of small libraries
eliminates ffpe samples, and also eliminates samples with less data.

Visualizing these data also informs the 'Analysis Stage 2' decision to
use VOOM for the differential expression analysis, which is useful were
library size variance is high.

```{r, out.width="80%"}

lib_size_cut = 3.7*10^7   # size cutoff used to filter samples for a subsequent query below

(p.ls <- dge[[2]] %>%
  as_tibble(rownames = "library") %>%
  ggplot(., aes(fct_reorder(library, lib.size), lib.size)) +
  geom_bar(stat = "identity", fun = "mean", aes(fill = definition), 
           alpha = 0.8, show.legend = TRUE) +
  labs(title = "Distribution of library sizes", 
       x = "Index of 612 KIRC libraries",
       y = "Transcriptome library size",
       fill = NULL) +
  theme_pubr() +
  ggplot2::scale_fill_viridis_d(option = "C") +
  geom_hline(yintercept = lib_size_cut, 
             linetype = "dashed", 
             size = 1, 
             color = "gray80",
             show.legend = TRUE) +
    theme(legend.position = "top", 
          axis.text.x = element_blank(),
          axis.ticks.x = element_blank()) +
    scale_linetype_manual(name = paste("Filter line -", lib_size_cut))
  )

larger_libraries <- dge[[2]] %>%     # list of barcodes for larger libs
  dplyr::filter(lib.size > lib_size_cut) %>% 
  rownames()

ggsave("FIGS KIRC prog sig_2/library_size_dist.tiff", p.ls, width = 6, height = 4, dpi = 300)
```

Scale normalization of the RNAseq read counts using TMM (trimmed mean of
M component) normalization method.

```{r, fig.width = 10 }
dge <- calcNormFactors(dge)
```

Plot library normalization factors

```{r, out.width="80%"}
(p.norm <- dge[[2]] %>%
  as_tibble(rownames = "library") %>%
  ggplot(., aes(fct_reorder(library, norm.factors), norm.factors)) +
  geom_bar(stat = "identity", fun = "mean", 
           aes(fill = ifelse(library %in% larger_libraries, "large", "small")), 
           show.legend = TRUE) +
  labs(title = "Determining whether variation in \nnormalization factors is great enough \nto require VOOM", 
       x = "Index of 612 KIRC libraries",
       y = "Normalization factor",
       subtitle = paste("max / min of normalization factors", round(max(dge[[2]]$norm.factors)/min(dge[[2]]$norm.factors) , 3), "fold.\n VOOM is needed." ),
       fill = "Library size") +
  theme_pubr() +
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        legend.position = "top") +
  ggplot2::scale_fill_viridis_d(option = "C")
)

ggsave("FIGS KIRC prog sig_2/normalization_factors.tiff", p.norm, width = 6, height = 4, dpi = 300)

```

Plot library expression distributions

```{r, out.width="100%", fig.width= 18}
(p.ld <- dge[[1]] %>%
  as_tibble(rownames = "geneID") %>%
  pivot_longer(-geneID,
               names_to = "library",
               values_to = "expression") %>%
  ggplot(., aes(`library`, log(`expression`, 10))) +
  geom_boxplot(outlier.colour="black",
               outlier.shape=16,
               outlier.size=2,
               notch=TRUE,
               alpha = 0.5,
               aes(fill = ifelse(library %in% larger_libraries, "large lib.", "small lib.")), show.legend = TRUE) +   # fill = library
  labs(title = "Library Distribution",
       x = "Index of all libraries (NAT and KIRC tumor)",
       y = "Log10 library size",
       fill = "Libraries retained = blue",
       subtitle = paste("max / min of lib.size", round(max(dge[[2]]$lib.size)/min(dge[[2]]$lib.size), 3), "fold" )) +
  theme(#axis.text.x = element_text(angle = 45, hjust = 1)
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        legend.position = "top") +
  ggplot2::scale_fill_viridis_d(option = "C"))

ggsave("FIGS KIRC prog sig_2/library_distribution.tiff", p.ld, width = 15, height = 4, dpi = 300)
```

<br><br>

------------------------------------------------------------------------

<br><br>

### Stage 2 - Differential expression analysis (tumor vs NAT)

Re-do GDC query, preparation, & normalization using a new filtered
sample list (samples with larger library sizes). Then use a Kaplan-Meier
analysis and COXPH model to find genes the predict survival.

So far...

A GDC query and download was done to visualize the metadata, identifing
potential confounders that should be corrected for in the final
voom/limma design matrix below. The library sizes and normalization
factors were plotted to find outliers, and a sample list excluding very
small libraries was created.

Next...

The GDC query will be set up again using the filtered sample list.
Design matrix and normalization factors are computed again to ensure
compatibility with the new query.

<br><br>

#### 2.1 GDC query of QC'ed library list

```{r, message = FALSE, results='hide', fig.keep='all'}

query_filtered = GDCquery(
  project = "TCGA-KIRC",
  data.category = "Transcriptome Profiling", # parameter enforced by GDCquery
  experimental.strategy = "RNA-Seq",
  workflow.type = "STAR - Counts",
  sample.type = c("Primary Tumor", "Solid Tissue Normal"),
  access = "open",
  legacy = FALSE, 
  barcode = larger_libraries                 # the filter "larger_librararies" is now added in an otherwise identical query
  )

# Re_create expression matrix
KIRC_data <- GDCprepare(query_filtered,
                        directory = target_directory)

clinical <- KIRC_data@colData %>% as_tibble %>%
  mutate(ajcc_pathologic_stage = case_when(  # make cancer stage data a continuous & quantitative variable
    ajcc_pathologic_stage == "Stage I" ~ 1,
    ajcc_pathologic_stage == "Stage II" ~ 2,
    ajcc_pathologic_stage == "Stage III" ~ 3,
    ajcc_pathologic_stage == "Stage IV" ~ 4,
    TRUE ~ 2
))  

group <- "definition"
comparison_groups <- KIRC_data@colData[[group]] %>% unique                   # set comparison groups
design_factor = colData(KIRC_data)[, group, drop=T]                          # choose design factor
design_factor <- factor(
    x = design_factor,
    levels=c("Solid Tissue Normal", "Primary solid Tumor")                   # Set first item here is the reference
)

design <- model.matrix(~ design_factor + gender + ajcc_pathologic_stage, clinical)

dge = DGEList(counts=assay(KIRC_data),  # CREATE S4 OBJECT
              samples=colData(KIRC_data),
              genes=as.data.frame(rowData(KIRC_data)))

keep <- filterByExpr(dge, design)       # output is a gene list

dge <- dge[keep,,keep.lib.sizes=FALSE]  # REMOVE LOW EXPRESSED GENES

dge <- calcNormFactors(dge)             # NORMALIZE

#  dge$counts %>% dim
```

<br><br>

#### 2.2 Limma/VOOM DEA

Voom does a transformation & differential expression calculation. VOOM
was chosen over limma-trend because the library sizes are variable
between samples (\>3-fold differences). The VOOM transformation is
applied to the normalized & filtered DGEList object 'dge'.

In the voom plot below, each point is one gene, with x axis showing the
avg expression in log2 cpm reads and y axis is the square-root of the
residual stdev of expression in log2 counts per million reads.

Voom transformation uses the experiment design matrix, and produces an
EList (S4) object

```{r}
voomObj <- voom(dge, design, plot=FALSE)
```

After VOOM, the usual LIMMA pipelines can be run, starting w fitting a
model

```{r}
fit <- lmFit(voomObj, design) # fit a linear model to the array
fit <- eBayes(fit) # ebays adds to the linear model. Given a linear model fit, eBayes computes moderated t stats, moderated f-stat, and log-odds of differential expression by empirical Bayes moderation of the standard errors toward a common value.
```

Derive & plot top significantly (BH FDR10) differentially expressed
genes (DEG) from the limma model.

```{r, out.width="80%"}
top_gene_count = 30000  # nrow(fit)   # look at all fitted genes

top_table <- topTable(fit, coef = ncol(design), adjust.method = "BH", n = top_gene_count)

p.top_table <- top_table %>% 
  as_tibble(rownames = "gene_names") %>%
  dplyr::filter(-log(adj.P.Val, 10) > 1) %>%
  ggplot(., aes(logFC, -log(adj.P.Val, 10))) +
  geom_pointdensity(alpha = 0.6, show.legend = FALSE) +
  geom_text_repel(aes(label = ifelse(-log(adj.P.Val,10) > 7, gene_name, "")), max.iter = 3000) +
  labs(title = "Top DEG \'top table\' method",
       x = "Log2FC (KIRC tumor / NAT)",
       y = "-Log10 FDR") +
  ggplot2::scale_color_viridis_c(option = "H", direction = -1) +
    scale_y_continuous(limits = c(0, 16), breaks = seq(0,16, 2), expand = expansion(mult = c(0, 0.1))) +
  scale_x_continuous(limits = c(-1,1), breaks = seq(-1,1, 0.5))

#  Or, to give more weight to fold changes in the ranking, one could say

fit <- treat(fit, lfc=log2(1.05))

top_treat <- topTreat(fit, coef=ncol(design), n = top_gene_count)

p.top_treat <- top_treat %>% 
  as_tibble(rownames = "gene_names") %>%
  dplyr::filter(-log(adj.P.Val, 10) > 1) %>%
  ggplot(., aes(logFC, -log(adj.P.Val,10))) +
    geom_pointdensity(alpha = 0.6, show.legend = FALSE) +
    geom_text_repel(aes(label = ifelse(-log(adj.P.Val,10) > 7, gene_name, "")), max.iter = 3000) +
    labs(title = "Top DEG \'top treat\' method",
         x = "Log2FC (KIRC tumor / NAT)",
         y = "-Log10 FDR") +
    ggplot2::scale_color_viridis_c(option = "H", direction = -1) +
    scale_y_continuous(limits = c(0, 16), breaks = seq(0,16, 2), expand = expansion(mult = c(0, 0.1))) +
    scale_x_continuous(limits = c(-1,1), breaks = seq(-1,1, 0.5))   

top_table %>% mutate(sig = ifelse(adj.P.Val < 0.1, "signif.", "non-signif.")) %>% .$sig %>% table
top_treat %>% mutate(sig = ifelse(adj.P.Val < 0.1, "signif.", "non-signif.")) %>% .$sig %>% table

(p_DEG_volcanos <- p.top_table + p.top_treat & 
  plot_annotation(caption = "Top DEG can be selected using a balance of p-values and fold-changes criteria. \nTop table emphasizes p values, while top treat emphasizes fold changes. \nTop genes above FDR10 are shown in both volcano plots."))

#The function treat() tests if the absolute log fold change is greater than the indicated value (log2(1.2)), rather than 0. 

ggsave("FIGS KIRC prog sig_2/volcanos.tiff", 
       p_DEG_volcanos, height = 6, width = 8)
```

<br><br>

```{r, include = FALSE}
limma_res <- list(
      voomObj=voomObj,       # normalized data
      fit=fit,               # fitted model and statistics
      topGenes=top_table %>% dplyr::filter(adj.P.Val < 0.1)     # top most differentially expressed genes
    )

meta_data <- limma_res$voomObj$targets
## components of voom object
# limma_res$voomObj[[1]] %>% head(., 5)             # gene name conversion
# limma_res$voomObj[[2]] %>% head(., 5)             # metadata
# limma_res$voomObj[[3]] %>% head(., 5)             # series matrix
```

<br><br>

Tabulate n values for Limma differential expression analysis above.

```{r, fig.width = 4, fig.height = 4, out.width="60%"}
(t_sampl_nvals <- meta_data %>%
  group_by(sample_type) %>%
  dplyr::count() %>%
  kbl(caption = "No. RNAseq samples after filtering.") %>%
  kable_paper("hover", full_width = F))

kableExtra::save_kable(t_sampl_nvals, 
                       file = "FIGS KIRC prog sig_2/GDC_sampl_nvals.html")

```

<br><br>

Plot top 20 DEG (derived comparing tumor vs. NAT):

```{r, fig.width = 9, fig.height = 10}

top_20_DEG <- top_table %>% 
  dplyr::arrange(`adj.P.Val`) %>% 
  head(., 20) %>% 
  dplyr::select(gene_id) %>% deframe

gene_vln <- function(gene) {
  gene_of_interest <- gene
  ext_gene_name <- limma_res$voomObj[[1]][gene_of_interest, "gene_name"]
  
  gene_meta <- t(limma_res$voomObj$E) %>%
    as_tibble(rownames = "barcode") %>%
    .[, c("barcode", gene_of_interest)] %>%
    left_join(., meta_data, by = "barcode") %>%
    dplyr::mutate(definition_rename = case_when(definition == "Primary solid Tumor" ~ "Tumor",
                                                definition == "Solid Tissue Normal" ~ "NAT",
                                                TRUE ~ "_"))
  
  gene_meta %>% 
    ggplot(., aes(definition_rename, .data[[gene_of_interest]])) +
    geom_violin(aes(fill = definition_rename), alpha = 0.5, show.legend = FALSE,
              draw_quantiles = c(0.25, 0.5, 0.75)) +
    stat_compare_means(method = "wilcox.test", 
              comparisons = list(c(1,2)), 
              aes(label=..p.adj..), 
              label = "p.signif") +
    labs(y = ext_gene_name, x = NULL) +
    scale_fill_aaas() +
    theme(axis.title.y=element_text(face="italic")) +
    scale_y_continuous(expand = expansion(mult = c(0, 0.1)))
}

(p_vln_topDEG <- map(top_20_DEG, ~ gene_vln(.x)) %>%
  gridExtra::marrangeGrob(., ncol = 4, nrow = 5))
```

<br><br>

#### 2.3 Sample stratification needed?

A useful prognostic signature will ideally mean a single signature that
predicts survival for all or most cases of a defined indication such as
KIRC. For clinicians, it is generally impractical to determine which
phenotypic subcluster a tumor transcriptome belongs to and then choose a
prognostic signature on the basis of such a classification.

Nonetheless, a single "one-size-fits-all" prognostic signature may not
be possible if there are multiple, distinct tumor sub-populations within
the KIRC classification. To explore this possibility, tumor sample
relationships are visualized here to determine whether multiple
prognostic signatures may be required.

Assessment of PCA plots below: Although there are clearly two
populations--NAT and tumor--the tumor samples do not clearly require
stratification. There is one, albeit phenotypically diverse tumor
population.

```{r, out.width="60%"}
plot_PCA_gg = function(vObj, condition_variable, pcs = c(1,2)) {
  group = factor(vObj$targets[, condition_variable])
  pca = prcomp(t(vObj$E))

  pca_plot <- ggplot(pca$x, aes(pca$x[,pcs[1]], pca$x[,pcs[2]])) +
         geom_point(aes(color = group), alpha = 0.6, size = 2) +
    theme(legend.position = "top") +
    labs(x = paste0("PC", pcs[1]),
         y = paste0("PC", pcs[2])) +
    scale_color_aaas()
  pca_plot
}

p.12 <- plot_PCA_gg(limma_res$voomObj, "definition", pcs = c(1,2))
p.34 <- plot_PCA_gg(limma_res$voomObj, "definition", pcs = c(3,4))

(p_PCA <- p.12 + p.34 & 
    labs(color = NULL) & 
    plot_annotation(title = "PCA of healthy tissue and renal tumor samples"))

ggsave("FIGS KIRC prog sig_2/PCA_plots.tiff", 
       p_PCA, width = 8, height = 4.5, dpi = 300)
```

The PCA plots above & scree plot below show that PC1 represents the
major component of difference between healthy and tumor tissues. PC's 1
thru 4 do not show distinct sub-populations within the KIRC tumor
samples, although there are a few phenotypic outliers. Based on this
observation, a single prognostic signature for all tumor samples will be
derived and tested for predictive power below.

```{r, out.width="60%"}

plot_scree <- function(scaled_matrix) {

  pca <- prcomp(t(scaled_matrix))
  pca_stats <- stats:::summary.prcomp(pca)$importance[2, ]

  pca_stats %>%
    as.matrix %>%
    t() %>%
    as_tibble() %>%
    pivot_longer(everything()) %>%
    .[1:10,] %>%    # only the first 10 prin. comps
    ggplot(., aes(x = fct_reorder(name, -value), y = value, label = value)) +
    geom_bar(alpha = 0.6, stat = "identity", fun = "mean", fill = "midnightblue") +
    scale_y_percent() +
    labs(x = NULL, y = "Fraction of variance", title = "Scree plot - first 10 PCs") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
}

(p_scree <- plot_scree(limma_res$voomObj$E) + plot_spacer())

ggsave("FIGS KIRC prog sig_2/Scree_plots.tiff", 
       p_scree, 
       width = 8, 
       height = 4)
```

Use a Spearman correlation heat map of the VOOM-normalized series matrix
to look at the prevalence of outliers and the number of tumor sample
populations. If multiple, clearly-distinct tumor populations are
present, sample stratification may be needed (i.e. comparing NAT to each
stratified tumor population).

A survival model will be constructed below to see whether a reliable
prognostic signature can be derived from an un-stratified set of tumor
samples.

```{r, out.width="60%"}
VO <- voomObj[[3]]

annotation_row = data.frame(Sample = factor(voomObj[[2]]$definition, ordered = TRUE))   # make row annotation
rownames(annotation_row) = voomObj[[2]]$barcode

cormat <- cor(VO, method = c("spearman"))   # correlation of all samples against each other

set.seed(123)

split <- annotation_row %>% 
  as_tibble(rownames = "gene_id") %>% 
  dplyr::filter(Sample == "Primary solid Tumor") %>% 
  dplyr::select(gene_id)

library(pals)
(p_heatmap <- pheatmap(cormat, 
         show_rownames = FALSE, 
         show_colnames = FALSE,
         show_column_dend = FALSE,
         color = viridis(60),
         cluster_rows = TRUE,
         annotation_row = annotation_row,
         row_split = annotation_row,
         annotation_colors = list(Sample = c("Primary solid Tumor"="royalblue4",
                                             "Solid Tissue Normal"="gray80")),   #  cadetblue3
         fontsize=12,
         main = "Correlation (rho) heat map of \nKIRC tumor & NAT"
         ) )

ggsave("FIGS KIRC prog sig_2/heatmap.tiff", p_heatmap, height = 5, width = 6.5, dpi = 1200)
```

<br><br>

------------------------------------------------------------------------

<br><br>

### Stage 3 - Predictive modeling

#### 3.1 Kaplan-Meier analysis

Abnormally expressed DEG (found by comparing tumor vs. NAT), were used
here to assess their predict value toward overall survival. The response
variable combines "days_to_last_follow_up" & "days_to_death", using
"vital_status" (death) as a censored variable. The output is a list of
genes with prognostic value, and these are used in the subsequent coxph
model.

KM Analysis input are tumor samples only, examining genes that are a) in
the top 10% of genes terms of high expression variance b) are
significantly differentially expressed between tumor and NAT
(Benjamini-Hochberg FDR10).

Data filtered prior to KM Analysis:

-   DEG \< FDR10

-   Top 10% of variance

-   Tumor samples only

```{r}
clin_df <- meta_data %>%
  as_tibble() %>%
  dplyr::filter(definition == "Primary solid Tumor") %>%    # filter on tumor sample only
  dplyr::select("barcode", "patient", "vital_status", "days_to_death", 
                "days_to_last_follow_up", "gender", "ajcc_pathologic_stage", "age_at_index") %>%
  mutate(deceased = ifelse(vital_status == "Dead", TRUE, FALSE))

clin_df$overall_survival = ifelse(clin_df$deceased, 
                                  clin_df$days_to_death,
                                  clin_df$days_to_last_follow_up)

# gives list of genes with high dispersion (top 10%)
d_mat <- varFilter(limma_res$voomObj$E, var.func=IQR, var.cutoff=0.9) %>% t() # Filter genes that are in the top 10% of variance/dispersion; note that limma_res is already filtered above DEG FDR10

gene_filter <- intersect(colnames(d_mat), top_table$gene_id)                  # filtering genes on tumor vs. NAT DEG @ FDR10
d_mat <- d_mat[, colnames(d_mat) %in% gene_filter]


```

```{r}
# divide patients in two groups, up and down regulated & then model whether survival is different between them.
pred_surv_pval <- function(x) {
  gene_id <- x
  median_value <- median(d_mat[,gene_id])
  
  d_meta_gene <- d_mat[,gene_id] %>% 
    as_tibble(rownames = "barcode") %>% 
    rename(., c("value" = "gene_value")) %>%
    inner_join(., clin_df, by = "barcode") %>%
    mutate(gene = ifelse(gene_value <= median_value, "DOWN", "UP"))
  
  # fit a survival model
  fit <- survfit(Surv(overall_survival, deceased) ~ gene, data=d_meta_gene)
  
  # extract the survival p-value
  pval = surv_pvalue(fit, data=d_meta_gene, method = "S1")$pval
  pval
}

  # KM Analysis output
surv_pvals <- map_dfr(colnames(d_mat),                      
                      ~ tibble(pval = pred_surv_pval(.x), 
                               ENSG = paste(.x))) %>% 
  dplyr::mutate(FDR = p.adjust(pval, method = "BH")) %>%
  dplyr::mutate(signif = ifelse(FDR <= 0.1, "signif", "non-signif"))

signif_pred_genes <- surv_pvals %>%                # Filter KM analysis on genes significant @ FDR10
  dplyr::filter(signif == "signif") %>%
  dplyr::select(ENSG) %>%
  deframe()
```

The KM analysis above generates a list of genes was found that
significantly predicted survival (BH FDR10) . However, p-values have
limited utility as sample numbers increase, so this list is refined
below include samples with high fold changes (cutoff was 2\^0.3219281 =
25% change).

```{r, message = FALSE}

fold_change <- function(x) {
    d_meta_gene <- d_mat[, x] %>% 
        as_tibble(rownames = "barcode") %>% 
        inner_join(., clin_df, by = "barcode") %>%
        mutate(gene = ifelse(value <= median(value), "DOWN", "UP")) # stratify samples based on gene expression (above or below that gene's median expression)
    
    two_average <- function(x) { log( x[1]/x[2], 2) }
    two_average(c(2,4))
    
    F2C <- d_meta_gene %>%
      group_by(gene) %>% 
      summarize(average = mean(overall_survival)) %>%
      dplyr::select(average) %>%
      deframe() %>% 
      two_average() 
    F2C
}

df_fc <- map_dfr(signif_pred_genes, 
                 ~ tibble(FC_log2 = fold_change(.x),
                                    ENSG = paste(.x)))

# Genes that predict survival based on both Kaplan-meier p-value and 
genes_signif_hiFC <- df_fc %>%             # ENSG gene list < 10%FDR & >15% fold change
  dplyr::filter(FC_log2 > 0.3219281) %>%   # change to 25% using 0.3219281   # or use 15%  0.2016339
  dplyr::select(ENSG) %>% 
  deframe()

df_hifc <- df_fc %>% 
  dplyr::filter(ENSG %in% genes_signif_hiFC)

(t_KM_topgenes <- surv_pvals %>%
  dplyr::filter(signif == "signif" & ENSG %in% genes_signif_hiFC) %>%
  full_join(., df_hifc, by = "ENSG") %>%
  dplyr::select(ENSG, pval, FDR, FC_log2) %>%
  dplyr::arrange(pval) %>%
  kbl(caption = "Kaplan-Meier Analysis\nGenes Predicting survival") %>%
  kable_paper("hover", full_width = F) )

kableExtra::save_kable(t_KM_topgenes, file = "FIGS KIRC prog sig_2/table_kap_meier_topgenes.html")

```

<br><br>

#### 3.2 Cox Proportional Hazards Model & AUC assessment

In order to further refine the prognostic gene list, and to generate an
AUC for the genes that predict survival according to the Kaplan-Meier
Analysis, a COXPH model us used here. The genes with the most
significant hazard ratios are used to calculate an AUC over time.

```{r, fig.height = 5, fig.width = 5, out.width="40%"}
# gene names for genes that predict survival according to kaplan-meier analysis
final_signature <- dge[[3]] %>% dplyr::filter(gene_id %in% genes_signif_hiFC)

coxph_df <- d_mat[,final_signature$gene_id] %>% 
    as_tibble(rownames = "barcode") %>% 
    inner_join(., clin_df, by = "barcode") %>% space_fix()

# vars facet object gives a list of symbol variables for the coxph model below. must be coped & pasted
vars <- final_signature$gene_id
vars_facet <- paste0(paste0("`", vars), "`") 
vars_facet <- paste(as.character(vars_facet), collapse = "+") %>% noquote()
```

Input the Kaplan-Meier genes into a Cox Proportional Hazards (COXPH)
model, in order to visualize their hazard ratios and p values in a
volcano plot for the entire data set.

```{r, out.width="50%"}

cox <- coxph(Surv(overall_survival, deceased) ~ `ENSG00000073150.14`+`ENSG00000100288.20`+`ENSG00000104313.20`+`ENSG00000106327.13`+`ENSG00000119547.6`+`ENSG00000123838.11`+`ENSG00000154277.13`+`ENSG00000155269.12`+`ENSG00000169885.10`+`ENSG00000179698.14`+`ENSG00000181449.4`+`ENSG00000184363.10`+`ENSG00000197599.12`+`ENSG00000198768.11`+`ENSG00000205702.11`+`ENSG00000213658.12`+`ENSG00000225217.1`+`ENSG00000225783.8`+`ENSG00000228727.9`+`ENSG00000231748.1`+`ENSG00000234537.1`+`ENSG00000236017.8`+`ENSG00000236453.5`+`ENSG00000236499.2`+`ENSG00000236591.1`+`ENSG00000236871.7`+`ENSG00000250067.12`+`ENSG00000253317.1`+`ENSG00000267121.6`+`ENSG00000268649.5`+`ENSG00000271959.1`+`ENSG00000277883.1`+`ENSG00000287089.1`, data=coxph_df)

gene_name_key <- read_csv("gene_namestop43.csv") %>% dplyr::select(gene_id, gene_name) %>% space_fix()

df_coef_top <- summary(cox) %>% 
  .$coefficients %>% 
  as_tibble(rownames = "gene_id") %>%
  left_join(., gene_name_key, by = "gene_id")

(p_hazratio <- df_coef_top %>%
  ggplot(., aes(`exp(coef)`, `Pr(>|z|)`)) +
  geom_hline(yintercept = 0.05, 
             linetype = "dashed", 
             color = "gray60",
             alpha = 0.5) +
  geom_vline(xintercept = c(0.9, 1.1), 
             linetype = "dashed", 
             color = "gray60",
             alpha = 0.5) +
  geom_point(aes(color = ifelse(`Pr(>|z|)` < 0.05, "p < 0.05", "n.s.")), 
             size = 1.5, alpha = 0.8) +
  scale_y_reverse(limits = c(1, -0.1)) +
  labs(color = NULL, x = "Hazard ratio (mortality)", y = "p-val.", title = "Top predictive genes - COXPH Model using the top 49 genes \nfrom the univariate KM analysis") +
  theme(legend.position = "top") +
  scale_color_manual(values = c("dodgerblue4", "deeppink3")) +
  geom_text_repel(aes(label = ifelse(`Pr(>|z|)` < 0.05, gene_name, "")), max.iter = 100000) +
  scale_x_continuous(n.breaks = 8) +
  theme_pubr())

ggsave("FIGS KIRC prog sig_2/Hazardratio_volcano.tiff", p_hazratio, width = 5, height = 5)

```

Determine AUC using genes with absolute hazard ratio \>=10%

<https://datascienceplus.com/time-dependent-roc-for-survival-prediction-models-in-r/>

Derive an AUC from only significant genes (from COXPH model for all
data). One of the 7 genes, *IP6K3*, is removed here as it is does not
have significant predictive power in the more targeted model (6 genes)
below. Above, the top genes are filtered above are further refined to a
smaller panel that would be feasible in a clinical setting (i.e. 5-10
gene transcripts).

```{r}

coxph_df_nam <- coxph_df %>%
  rename(., c("ENSG00000119547.6"= "ONECUT2"),
         c("ENSG00000155269.12" = "GPR78"),
         c("ENSG00000213658.12" = "LAT"),
         c("ENSG00000228727.9" = "SAPCD1"),
         c("ENSG00000236453.5" = "Lnc-GNG11-3"),
         c("ENSG00000236499.2" = "LINC00896"),
         c("ENSG00000287089.1" = "IP6K3") ) %>%
  space_fix()

top_seven_genes <- df_coef_top %>%   
  dplyr::filter(`Pr(>|z|)` < 0.05) %>%  #   abs(`exp(coef)` - 1) > 1 & 
  dplyr::select(gene_id, gene_name) %>% space_fix()

top_seven_genes_name <- top_seven_genes %>% .$gene_name
top_seven_genes_id <- top_seven_genes %>% .$gene_id


vars <- top_seven_genes_name
vars_facet <- paste0(paste0("`", vars), "`") 
#vars_facet <- as.formula(paste("~", paste(as.character(vars_facet), collapse = "+")) )
vars_facet <- paste(as.character(vars_facet), collapse = "+") %>% noquote()
```

Testing and plotting a AUC over time for the top 7 prognostic genes.

```{r, out.width="60%"}
set.seed(123) 
sample = sample.split(coxph_df_nam$barcode, SplitRatio = .75)

train = subset(coxph_df_nam, sample == TRUE)
test  = subset(coxph_df_nam, sample == FALSE)

train.fit <- coxph(Surv(overall_survival, deceased) ~ `ONECUT2`+`GPR78`+`LAT`+`SAPCD1`+`Lnc-GNG11-3`+`LINC00896`+`IP6K3`,
                   data=train, 
                   x=TRUE, 
                   y=TRUE, 
                   method="breslow")


set.seed(234)
doParallel::registerDoParallel()

lp <- predict(train.fit)                                    # predicted values for train
lpnew <- predict(train.fit, newdata=test)                   # predicted values for test
Surv.rsp <- Surv(train$overall_survival, train$deceased)    # train values
Surv.rsp.new <- Surv(test$overall_survival, test$deceased)  # test values
times <- seq(10, 4500, 10)
AUC_CD7 <- AUC.cd(Surv.rsp, Surv.rsp.new, lp, lpnew, times) 

unregister <- function() {
  env <- foreach:::.foreachGlobals
  rm(list=ls(name=env), pos=env)
}
unregister()

(p.auc <- as.data.frame(AUC_CD7[1:2]) %>%
  ggplot(aes(times/365, auc)) +
  geom_line(color = "dodgerblue4") +
  scale_y_continuous(limits=c(0.5, 0.9)) +
  geom_hline(yintercept = 0.5, 
             linetype = "dashed", 
             color = "deeppink4") +
  labs(x = "Years from Index", 
       y = "AUC", 
       title = "Proportion of time model accurately predicts survival", 
       subtitle = paste(paste(top_seven_genes_name[1:4], collapse = ", "), "\n", paste(top_seven_genes_name[5:7], collapse = ", "))) +
  theme(plot.subtitle = element_text(face = "italic")) )

ggsave("FIGS KIRC prog sig_2/AUC over time.tiff", plot = p.auc, height = 3, width = 5)

```

Tabulate coefficients for the 7-gene COXPH model

```{r, out.width="60%"}

coxph_seven_genes <- coxph(Surv(overall_survival, deceased) ~ `ONECUT2`+`GPR78`+`LAT`+`SAPCD1`+`Lnc-GNG11-3`+`LINC00896`+`IP6K3`,
                   data=coxph_df_nam, 
                   x=TRUE, 
                   y=TRUE, 
                   method="breslow")

(p_haz_ratios <- ggforest(coxph_seven_genes, data = coxph_df_nam, main = "Hazard ratios for COXPH model"))

ggsave("FIGS KIRC prog sig_2/Haz_ratios_7gene.tiff", p_haz_ratios, height = 2.8, width = 6, dpi = 600)

t_coxph_coefs <- summary(coxph_seven_genes) %>% 
  .$coefficients %>% 
  as_tibble(rownames = "gene_id") %>%
  left_join(., gene_name_key, by = "gene_id") %>%
  kbl(caption = "seven-gene COXPH model Coefficients") %>%
  kable_paper("hover", full_width = F)
  
kableExtra::save_kable(t_coxph_coefs, 
                       file = "FIGS KIRC prog sig_2/coxph_coefs.html")
```

Kaplan-Meier univariate survival curves showing the top 7 predictive
genes from the COXPH model above.

```{r, fig.width = 8, fig.height = 9, out.width="60%"}

clin_df$overall_survival_yrs <- clin_df$overall_survival/365

plot_surv_wpval <- function(x, y) {
  gene_id <- x
  median_value <- median(d_mat[,gene_id])
  
  gene_name <- dge[[3]] %>% 
    dplyr::filter(gene_id == x) %>%
    dplyr::select(gene_name) %>% deframe
  
  d_meta_gene <- d_mat[,gene_id] %>% 
    as_tibble(rownames = "barcode") %>% 
    rename(., c("value" = "gene_value")) %>%
    inner_join(., clin_df, by = "barcode") %>%
    dplyr::mutate(gene = ifelse(gene_value <= median_value, 
                                "\u2264 median", "> median"))

  fit <- survfit(Surv(overall_survival_yrs, deceased) ~ gene, data=d_meta_gene)

  ggsurvplot(fit, data=d_meta_gene, pval=T, 
             palette = "lancet",
             risk.table=F,
             xlab = "Years after index",
             title = y,
             conf.int = TRUE)
}

p_final_signature <- map2(.x = top_seven_genes[[1]],
                          .y = top_seven_genes[[2]],
                        ~ plot_surv_wpval(.x, .y) +
  theme_survminer(font.main = c(16, "italic", "black"),
                  legend.title="") ) %>% 
  arrange_ggsurvplots(
    .,
    title = NA,
    ncol = 2,
    nrow = 2
    )

ggsave("FIGS KIRC prog sig_2/Prognostic signature KM curves.pdf", 
       p_final_signature, 
       height = 8, 
       width = 6, 
       dpi = 300)

saveRDS(coxph_df, "coxph_df.rds")
```

#### 3.3 Incorporate new variables into the COXPH model: age_at_index, gender, & ajcc_pathologic stage into the model

```{r}
coxph_df <- readRDS("coxph_df.rds", refhook = NULL)

gene_name_key <- read_csv("gene_namestop43.csv") %>% dplyr::select(gene_id, gene_name) %>% space_fix()
```

Input here into the new cox model are the top genes from the KM analysis
above, but this time, incorporating the three clinical vars to see if
they enhance the model.

```{r, out.width="70%"}

cox <- coxph(Surv(overall_survival, deceased) ~ (age_at_index + gender + ajcc_pathologic_stage) * `ENSG00000073150.14`+`ENSG00000100288.20`+`ENSG00000104313.20`+`ENSG00000106327.13`+`ENSG00000119547.6`+`ENSG00000123838.11`+`ENSG00000154277.13`+`ENSG00000155269.12`+`ENSG00000169885.10`+`ENSG00000179698.14`+`ENSG00000181449.4`+`ENSG00000184363.10`+`ENSG00000197599.12`+`ENSG00000198768.11`+`ENSG00000205702.11`+`ENSG00000213658.12`+`ENSG00000225217.1`+`ENSG00000225783.8`+`ENSG00000228727.9`+`ENSG00000231748.1`+`ENSG00000234537.1`+`ENSG00000236017.8`+`ENSG00000236453.5`+`ENSG00000236499.2`+`ENSG00000236591.1`+`ENSG00000236871.7`+`ENSG00000250067.12`+`ENSG00000253317.1`+`ENSG00000267121.6`+`ENSG00000268649.5`+`ENSG00000271959.1`+`ENSG00000277883.1`+`ENSG00000287089.1`, data=coxph_df)

df_coef_top <- summary(cox) %>% 
  .$coefficients %>% 
  as_tibble(rownames = "gene_id") %>%
  left_join(., gene_name_key, by = "gene_id") %>% space_fix()

(p_hazratio <- df_coef_top %>%
  ggplot(., aes(`exp(coef)`, `Pr(>|z|)`)) +
  geom_hline(yintercept = 0.05, 
             linetype = "dashed", 
             color = "gray60",
             alpha = 0.5) +
  geom_vline(xintercept = c(0.9, 1.1), 
             linetype = "dashed", 
             color = "gray60",
             alpha = 0.5) +
  geom_point(aes(color = ifelse(`Pr(>|z|)` < 0.05, "p < 0.05", "n.s.")), 
             size = 1.5, alpha = 0.8) +
  scale_y_reverse(limits = c(1, -0.1)) +
  labs(color = NULL, x = "Hazard ratio (mortality)", y = "p-val.", title = "Top predictive genes - COXPH Model using the top 49 genes \nfrom the univariate KM analysis") +
  theme(legend.position = "top") +
  scale_color_manual(values = c("dodgerblue4", "deeppink3")) +
  geom_text_repel(aes(label = ifelse(`Pr(>|z|)` < 0.05, gene_name, "")), max.iter = 500) +
  #scale_x_continuous(n.breaks = 8, limits = c(-0.5, 5.5), breaks = seq(0.5, 5.5, 0.5)) +
  scale_x_log10() +
  theme_pubr())

ggsave("FIGS KIRC prog sig_2/Hazardratio_volcano_3clin.tiff", 
       p_hazratio, width = 5, height = 5)
```

```{r, out.width="70%"}

coxph_df_namx <- coxph_df %>%
  rename(., c("ENSG00000119547.6"= "ONECUT2",
         "ENSG00000155269.12" = "GPR78",
         "ENSG00000213658.12" = "LAT",
         "ENSG00000228727.9" = "SAPCD1",
         "ENSG00000236453.5" = "Lnc-GNG11-3",
         "ENSG00000287089.1" = "IP6K3",
         "ENSG00000205702.11" =	"CYP2D7") ) %>%
  space_fix()

top_seven_genes <- df_coef_top %>%   
  dplyr::filter(`Pr(>|z|)` < 0.05) %>%  #   abs(`exp(coef)` - 1) > 1 & 
  dplyr::select(gene_id, gene_name) %>% 
  space_fix() 

top_seven_genes_name <- top_seven_genes %>% .$gene_name
top_seven_genes_id <- top_seven_genes %>% .$gene_id

vars <- top_seven_genes_name
vars_facet <- paste0(paste0("`", vars), "`") 
#vars_facet <- as.formula(paste("~", paste(as.character(vars_facet), collapse = "+")) )
vars_facet <- paste(as.character(vars_facet), collapse = "+") %>% noquote()
```

Tabulate coefficients for the 7-gene COXPH model along with age, gender,
& ajcc pathologic stage.

```{r, out.width="70%"}
coxph_seven_genes_x <- coxph(Surv(overall_survival, deceased) ~ (age_at_index + gender + ajcc_pathologic_stage) * `ONECUT2`+`GPR78`+`CYP2D7`+`LAT`+`SAPCD1`+`Lnc-GNG11-3`+`IP6K3`,
                   data=coxph_df_namx, 
                   x=TRUE, 
                   y=TRUE, 
                   method="breslow")

t_coxph_coefs_extras <- summary(coxph_seven_genes_x) %>% 
  .$coefficients %>% 
  as_tibble(rownames = "gene_id") %>%
  left_join(., gene_name_key, by = "gene_id") %>%
  kbl(caption = "seven-gene COXPH model Coefficients") %>%
  kable_paper("hover", full_width = F)

kableExtra::save_kable(t_coxph_coefs_extras, file = "FIGS KIRC prog sig_2/coxph_coefs_extras.html")

(p_haz_ratios_extra <- ggforest(coxph_seven_genes_x, data = coxph_df_namx, 
                                main = "Hazard Ratios"))

ggsave("FIGS KIRC prog sig_2/Haz_ratios_7gene_clinvars.tiff", p_haz_ratios_extra, height = 4.5, width = 8)

```


### Discussion

A prognostic signature of six transcripts was derived here. This
signature predicts survival with the Chambless and Diao's estimator of
cumulative/dynamic AUC = 0.8 for the COXPH model. This predictive
capacity that remains durable over 12 years after index (sample
collection time). The addition of clinical variables to the model--including tumor stage & patient age at index--does add moderate predictive capacity to the model, but the advantage is only slight. These prognostic transcripts include transcripts known to contribute to tumor
progression, including NFùõãB signaling (*LAT*) , HIF-1‚ç∫ signaling
(*ONECUT2*), as well as inositol hexakisphosphate kinase 3 (*IP6K3*) &
glucose-regulated protein 78 (*GPR78*), both of which have a previously
defined negative prognostic value in renal cancer. Prognostic genes also
included cytochrome P450, *CYP2D7* consistent therapeutic resistance
predicting survival. Expression of these genes provide valuable
prognostic information for KIRC patients.

<br><br>

------------------------------------------------------------------------

References:

*ONECUT2* - Accelerates tumor progression in gastric, CRC, prostate,
hepatocellular cancers. Androgen resistance and HIF1alpha signaling
pathways.

<https://www.nature.com/articles/s41467-018-08133-6>

*IP6K3* - prognostic in renal cancer

<https://www.proteinatlas.org/ENSG00000161896-IP6K3>

*LAT* - NFkB signalling; immune signalling; MHC and TCR signaling

<https://www.proteinatlas.org/ENSG00000213658-LAT/pathology/renal+cancer>

*GPR78* - prognostic in renal cancer

see kaplan-meier plot
<https://www.proteinatlas.org/ENSG00000155269-GPR78/pathology/renal+cancer>

*LINC00896*

<https://pubmed.ncbi.nlm.nih.gov/34790382/>

<https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8576217/>

*SAPCD1* - DNA mismatch repair; prognostic in cancer

<https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7391198/>

*CYP2D6* - drug resistance / detoxification
<https://www.frontiersin.org/articles/10.3389/fphar.2010.00121/full>

Long-non-coding RNA's show prognostic value in KIRC

<https://www.frontiersin.org/article/10.3389/fonc.2020.01430>
------------------------------------------------------------------------
